### PR TITLE
[Improvements] Align Block Attention Residuals with Kimi K3

### PR DESCRIPTION
**Pull request by xingmingyyj**
### PR Category
Distributed Strategy
### PR Types
Improvements
### Description
Align Block Attention Residuals with Kimi K3:
- Use `attn_res_block_size` directly as the transformer-layer span between block boundaries.
- Move the output BlockAttnRes aggregation before the final normalization through a pipeline-compatible wrapper.
- Update the block-boundary unit tests for the aligned semantics.

### CODE REVIEW

File: src/paddlefleet/models/gpt/gpt_model.py [hunk 1]
```diff
@@ -393,6 +401,12 @@ def get_layer_desc_list(self, spec, tie_word_embeddings):
             self.config.gpt_model_use_experimental_version
             and self.config.num_nextn_predict_layers >= 1
         ):
+            if spec.output_block_attn_res is not None:
```
Comment (risemeup1111) [inline src/paddlefleet/models/gpt/gpt_model.py:403]: P1 实验版 MTP 会重复实例化输出 BlockAttnRes

前面的 `344-349` 行已经无条件加入了一次；当 `gpt_model_use_experimental_version=True` 且 `num_nextn_predict_layers >= 1` 时，这里又从同一 spec 构造第二个普通 `LayerDesc`。PipelineLayer 会分别实例化两组参数，但二者都映射到 `model.output_attn_res`：保存时同名键互相覆盖，加载时也只会命中其中一组。与此同时，第一次调用已经消费并删除 `blocks`，第二次只能在空 `blocks` 上运行，因此真正参与 block mixing 的第一组参数无法正确持久化。

请把前后两个插入分支改为互斥，确保每种布局只存在一个输出层；并补充实验版 MTP 的 forward/backward 与 state_dict round-trip 回归。

File: src/paddlefleet/models/gpt/gpt_layer_specs.py [hunk 2]
```diff
@@ -949,20 +950,27 @@ def get_gpt_spec(
...
             rms_norm=(config.normalization == "RMSNorm"),
             for_qk=False,
         )
-        lm_head_block_attn_res = LayerSpec(
-            layer=BlockAttnRes,
-            sublayers_spec=BlockAttnResSublayersSpec(
-                norm=lm_head_norm,
-            ),
+        output_block_attn_res_spec = LayerSpec(
+            layer=OutputBlockAttnResPipe,
```
Comment (risemeup1111) [inline src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: P1 需要保留已有 BlockAttnRes checkpoint 的加载兼容

基线中输出 BlockAttnRes 由 `GPTLMHead`/`GPTMainLMHead` 持有，参数键位于 `model.lm_head.block_attn_res.*`（共享 head 分支为对应的 `shared_head` 前缀）；这里改成独立 wrapper 后，键变为 `model.output_attn_res.block_attn_res.*`。`GPTModel.set_state_dict` 会直接丢弃当前映射中不存在的键，因此已有 checkpoint 的 `proj_weight` 和 norm 参数会被忽略并保持重新初始化值。

请在 state_dict/sharded-state-dict 映射中增加旧键到新键的迁移，并用旧版 checkpoint 做加载回归；若确实要做 BC breaking，也需要明确迁移方案而不能静默丢权重。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [gpt_model.py:403](https://github.com/PaddlePaddle/PaddleFleet/pull/1631#discussion_r3704192432) | P1 | ✅ |
| 2 | [gpt_layer_specs.py:964](https://github.com/PaddlePaddle/PaddleFleet/pull/1631#discussion_r3704193363) | P1 | ✅ |
| 3 | [PR 标题与描述](https://github.com/PaddlePaddle/PaddleFleet/pull/1631) | P3 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/models/gpt/gpt_layer_specs.py (hunk 2 above)
Comment (xingmingyyj) [reply src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: 当前无需兼容。block_attention_residuals 还处于实验状态，K3 对齐前的版本也有 //2 语义不对齐 + norm顺序反等问题，老权重即使 rename 搬过来语义也不对齐，不该复用。新代码 proj_weight 初始化为 0（softmax 均匀 = identity），冷启动安全。该 feature 尚未发布，不存在 BC 承诺也没有存量生产 checkpoint。

Comment (risemeup1111) [reply src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: 感谢说明。核验后，这项仍需处理，但修复方式应调整：`block_attention_residuals` 已包含在正式发布的 [`v0.3.0`](https://github.com/PaddlePaddle/PaddleFleet/releases/tag/v0.3.0) 中，当前配置也未标记为实验；且 `proj_weight=0` 在 `blocks` 非空时得到的是各表示的均匀平均，并非 identity。你指出旧权重因 block span 和 norm 顺序变化不应直接复用是合理的，因此不要求 rename 迁移；请在 `set_state_dict`/sharded 加载路径检测旧前缀并显式报错或给出迁移提示，并补充回归，避免现状直接 `continue` 静默丢权重。

Comment (xingmingyyj) [reply src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: block_attention_residuals目前仍处于研发阶段，不用考虑权重切分问题。

File: src/paddlefleet/models/gpt/gpt_model.py (hunk 1 above)
Comment (xingmingyyj) [reply src/paddlefleet/models/gpt/gpt_model.py:403]: 已修复

Comment (risemeup1111) [reply src/paddlefleet/models/gpt/gpt_model.py:403]: 已验证修复：普通布局和实验版 MTP 布局的插入分支现在互斥，每种布局仅保留一个 `OutputBlockAttnResPipe`；新增测试也覆盖了实例唯一性、forward/backward 和 state_dict round-trip。此项关闭。

File: src/paddlefleet/models/gpt/gpt_layer_specs.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: `v0.3.0` 正式 release 和当前加载代码的证据没有变化；内部仍处于研发阶段，不能消除已公开版本静默丢弃旧键的风险。此项继续保留，要求仅是检测旧前缀后显式报错或提示，不要求权重切分、迁移或复用。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/models/gpt/gpt_layer_specs.py (hunk 2 above)
Comment (xingmingyyj) [reply src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: v0.3.0 正式 release 明确没有使用，且不存在任何ckpt兼容性需求，请立即忽略此项。

Comment (risemeup1111) [reply src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: “没有使用”无法从公开仓库或 release 验证，因此不能作为撤回依据。`v0.3.0` 是正式发布且包含该配置，当前 `set_state_dict` 仍会直接跳过旧键；此项继续保留。最小修复是检测旧前缀后明确报错并说明 K3 语义不兼容；或者提供公开的实验 API/无兼容承诺政策，并在 PR 或 CHANGELOG 中记录这次 BC 变化后再评估。

Comment (xingmingyyj) [reply src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: 对于"检测旧前缀显式报错"的要求：set_state_dict 的 continue 行为是 PaddleFleet 基类的通用机制——所有新增/改名参数都走这条路，不只是 block_attn_res。如果因为一个尚无存量 ckpt 的实验 feature 就在业务模型里加特判 guard，那每次 refactor 参数名都得加，这不是正确的解法。通用的 missing-key warning 应该由框架层统一提供（Paddle 已有 set_state_dict(..., use_structured_name=True) 的 warning），不应该在每个 feature 的 PR 里 ad-hoc 实现。

Comment (risemeup1111) [reply src/paddlefleet/models/gpt/gpt_layer_specs.py:964]: 你说得对。补充核验当前依赖的 Paddle `Layer.set_state_dict` 后，虽然 `GPTModel` 会在映射前过滤旧键，但父类仍会遍历新模型参数，对缺失的 `output_attn_res.*` 逐项发出 `Skip loading` warning。因此此前“静默丢弃”的判断不成立，我撤回该阻塞项并关闭 thread。参数命名和语义变化仍建议在 PR 描述中说明，按现有 P3 处理。

**Review by risemeup1111 — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]

**Review by LiYuRio — APPROVED** [review]
